### ISSUE TITLE
[Hold for #36378][$500] The member list of a public room shows inconsistent results

### ISSUE THREAD

**Issue by m-natarajan** [Expensify/App#37192]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.43-14
**Reproducible in staging?:**
**Reproducible in production?:**
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @quinthar 
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1708893734376029

## Action Performed:
Precondition: Have a public room, users with real accounts and anonymous users without email contact information
1. Go to the public room
2. Click on the header to see the details of the room
3. Look at the members count
4. Click on the members to see the members list
5. Go back to the details page and look at the members count

## Expected Result:
Should show the same number of the members 

## Actual Result:
It first showing the  correct number of people who have signed in with real accounts and when opening the members list and going back it includes anonymous user (who didn't provide any email contact information)

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [ ] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

![image (6)](https://github.com/Expensify/App/assets/38435837/e937e544-7f66-4401-bd9d-32c796c0c61f)

![image (7)](https://github.com/Expensify/App/assets/38435837/9d3eb575-fe43-4187-aa27-1eace0e36628)

https://github.com/Expensify/App/assets/38435837/011eb9d4-f4d4-4fd8-a055-12c389719476

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01a589af5940630bea</li>
        <li>Upwork Job ID: 1761932775026454528</li>
        <li>Last Price Increase: 2024-02-26</li>
    </ul>
</details>

**Comment by kadiealexander** [comment]
Reproduced in chrome.

**Comment by brunovjk** [comment]
I believe they are similar https://github.com/Expensify/App/issues/36378

**Comment by wildan-m** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

The member list of a public room shows inconsistent results

### What is the root cause of that problem?

We have not omitted any visible members without personal details on the report detail page.

### What changes do you think we should make in order to solve the problem?

We can filter to only include visible member with personalDetails in [getVisibleMemberIDs](https://github.com/Expensify/App/blob/89d5689978d3a50c86ae9c9da00dcff875eb9220/src/libs/ReportUtils.ts#L4684)

```
function getVisibleMemberIDs(report: OnyxEntry<Report>): number[] {
    if (!report) {
        return [];
    }
    const visibleChatMemberAccountIDs = report.visibleChatMemberAccountIDs ?? [];

    let visibleMemberIDs: number[] = [];
    report?.visibleChatMemberAccountIDs?.forEach((accountID) => {
        const details = allPersonalDetails?.[accountID];

        if (!details) {
            return;
        }

        visibleMemberIDs.push(accountID);
    });

    // Build participants list for IOU/expense reports
    if (isMoneyRequestReport(report)) {
        const onlyTruthyValues = [report.managerID, report.ownerAccountID, ...visibleChatMemberAccountIDs].filter(Boolean) as number[];
        const onlyUnique = [...new Set([...onlyTruthyValues])];
        return onlyUnique;
    }

    return visibleMemberIDs;}
```

then we can remove this code to avoid redundant code:

https://github.com/Expensify/App/blob/89d5689978d3a50c86ae9c9da00dcff875eb9220/src/pages/RoomMembersPage.tsx#L156-L161

Or we can do filtering per page like RoomMembersPage did inside ReportDetailsPage.tsx

### What alternative solutions did you explore? (Optional)

**Comment by Emrullah-Akbulak** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Member count in a public room's details shows inconsistency when the user enters the "RoomMembersPage" and goes back to the "ReportDetailsPage". 

### What is the root cause of that problem?

When user opens a room the client sends an API command called "OpenReport" which returns the report information that is used to display the room's details such as member count. However, when the user navigates to "RoomMembersPage" via clicking "Members" button, the client sends another command called "OpenRoomMembersPage". This new command's response returns new object which has the following properties "participantAccountIDs", "participants", "visibleChatMemberAccountIDs", then this fields gets merged with the existing report via Onyx. 

The problem is the participantAccountIDs fields between these requests "OpenReport" and "RoomMembersPage" returns different values and that causes "RoomMembersPage" command's response to override these 3 fields. When the fields get overwritten, the following code in "ReportDetailsPage"

```ts
const participants = useMemo(() => {
        console.log("I have RUN" ,report);
        return ReportUtils.getVisibleMemberIDs(report)
    }, [report]);
```
updates the participant count with the new overwritten values due to report object being updated by "OpenRoomMembersPage" command. Hence the member count changes when the user navigates & go backs from the "RoomMembersPage".

Moreover, the code that calls the "OpenRoomMembersPage" command is located at RoomMembersPage and wrapped in useEffect hook that runs only once the component mounts.

```ts
    const getRoomMembers = useCallback(() => {
        if (!report) {
            return;
        }
        Report.openRoomMembersPage(report.reportID);
        setDidLoadRoomMembers(true);
    }, [report]);

    useEffect(() => {
        getRoomMembers();
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [])
```

### What changes do you think we should make in order to solve the problem?

Assuming that "OpenReport" command returns the true value of the report. The "OpenRoomMembersPage" should either use the same filtration as the "OpenReport" command so that it filters and returns the same number of members to client.

### What alternative solutions did you explore? (Optional)

In my opinion, the "OpenRoomMembersPage" command can also be completely removed or altered since it permanently adds the "participantAccountIDs", "participants", "visibleChatMemberAccountIDs" fields to report object for no apparent reason other then being used in "RoomMembersPage" page . I haven't checked all the code that uses these fields but if these 3 fields are only going to be added once the user enter the "RoomMembersPage " page then they should get disposed and not permanently alter the report object when the user goes back.

**Comment by Emrullah-Akbulak** [comment]
Contributor details
Your Expensify account email: emrllhakb@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01565aae43960f7cbc

**Comment by kadiealexander** [comment]
Holding for https://github.com/Expensify/App/issues/36378!

**Comment by kadiealexander** [comment]
Still on hold.

**Comment by kadiealexander** [comment]
Same

**Comment by kadiealexander** [comment]
Still on hold.

**Comment by kadiealexander** [comment]
Asked for a retest [here](https://expensify.slack.com/archives/C9YU7BX5M/p1715655661790779).

**Comment by kbecciv** [comment]
QA team is unable to reproduce it on build 1.4.74.3

https://github.com/Expensify/App/assets/93399543/21fe43e5-89d9-4406-92da-297bf481ac62


https://github.com/Expensify/App/assets/93399543/4400dc04-58c1-4cd9-ab8c-439cf800c03b
